### PR TITLE
[SAIC-565] Network info retreiving optimization

### PR DESCRIPTION
**Pull request by miarmak**
Reduce OpenStack Neutron API calls in the `NeutronNetwork's`
`read_info` method. Use caching instead of multiple API calls.

Fix issue with keyword arguments in `utils/cache` module.

Add `get_all_quota` option to all configs.

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.base.network
cloudferrylib/base/network.py:25: [E0711(notimplemented-raised), Network.get_func_mac_address] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:24: [W0613(unused-argument), Network.get_func_mac_address] Unused argument 'instance'
cloudferrylib/base/network.py:28: [E0711(notimplemented-raised), Network.create_port] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'mac'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'tenant_id'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'sg_ids'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'ip'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'net_id'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'keep_ip'
cloudferrylib/base/network.py:31: [E0711(notimplemented-raised), Network.delete_port] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:30: [W0613(unused-argument), Network.delete_port] Unused argument 'port_id'
cloudferrylib/base/network.py:34: [E0711(notimplemented-raised), Network.check_existing_port] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:33: [W0613(unused-argument), Network.check_existing_port] Unused argument 'network_id'
cloudferrylib/base/network.py:33: [W0613(unused-argument), Network.check_existing_port] Unused argument 'mac'
cloudferrylib/base/network.py:37: [E0711(notimplemented-raised), Network.get_security_groups] NotImplemented raised - should raise NotImplementedError
************* Module tests.cloudferrylib.os.network.test_neutron
tests/cloudferrylib/os/network/test_neutron.py:48: [E1002(super-on-old-class), NeutronTestCase.setUp] Use of super on an old style class
tests/cloudferrylib/os/network/test_neutron.py:186: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:188: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:191: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:196: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:201: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:1476: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4
cloudferrylib/os/network/neutron.py:1486: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4
cloudferrylib/os/network/neutron.py:1488: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4
cloudferrylib/os/network/neutron.py:1489: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 8
cloudferrylib/os/network/neutron.py:1490: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 8
cloudferrylib/os/network/neutron.py:1491: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 8
cloudferrylib/os/network/neutron.py:1492: [W0311(bad-indentation), ] Bad indentation. Found 16 spaces, expected 12
cloudferrylib/os/network/neutron.py:1493: [W0311(bad-indentation), ] Bad indentation. Found 20 spaces, expected 16
cloudferrylib/os/network/neutron.py:173: [W0221(arguments-differ), NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:283: [W0621(redefined-outer-name), NeutronNetwork.get_network] Redefining name 'network' from outer scope (line 24)
cloudferrylib/os/network/neutron.py:759: [W0221(arguments-differ), NeutronNetwork.get_security_groups] Arguments number differs from overridden 'get_security_groups' method
cloudferrylib/os/network/neutron.py:863: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_vips] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:888: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_members] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:919: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_monitors] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:980: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_pools] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:1371: [W1504(unidiomatic-typecheck), NeutronNetwork.get_resource_hash] Using type() instead of isinstance() for a typecheck.
cloudferrylib/os/network/neutron.py:1378: [W1504(unidiomatic-typecheck), NeutronNetwork.get_resource_hash] Using type() instead of isinstance() for a typecheck.
cloudferrylib/os/network/neutron.py:1382: [W1504(unidiomatic-typecheck), NeutronNetwork.get_resource_hash] Using type() instead of isinstance() for a typecheck.

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/1440/

**Comment by MirantisWorkloadMobility** [comment]
Functional test success for Grizzly to Juno migration

**Comment by MirantisWorkloadMobility** [comment]
Functional test success for Grizzly to Icehouse migration

**Comment by Settis** [comment]
+1

**Comment by miarmak** [comment]
Updated only log level in the `cloudferrylib/os/network/neutron.py` from warning to debug.

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.base.network
cloudferrylib/base/network.py:25: [E0711(notimplemented-raised), Network.get_func_mac_address] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:24: [W0613(unused-argument), Network.get_func_mac_address] Unused argument 'instance'
cloudferrylib/base/network.py:28: [E0711(notimplemented-raised), Network.create_port] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'mac'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'tenant_id'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'sg_ids'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'ip'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'net_id'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'keep_ip'
cloudferrylib/base/network.py:31: [E0711(notimplemented-raised), Network.delete_port] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:30: [W0613(unused-argument), Network.delete_port] Unused argument 'port_id'
cloudferrylib/base/network.py:34: [E0711(notimplemented-raised), Network.check_existing_port] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:33: [W0613(unused-argument), Network.check_existing_port] Unused argument 'network_id'
cloudferrylib/base/network.py:33: [W0613(unused-argument), Network.check_existing_port] Unused argument 'mac'
cloudferrylib/base/network.py:37: [E0711(notimplemented-raised), Network.get_security_groups] NotImplemented raised - should raise NotImplementedError
************* Module tests.cloudferrylib.os.network.test_neutron
tests/cloudferrylib/os/network/test_neutron.py:48: [E1002(super-on-old-class), NeutronTestCase.setUp] Use of super on an old style class
tests/cloudferrylib/os/network/test_neutron.py:186: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:188: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:191: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:196: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:201: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:1478: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4
cloudferrylib/os/network/neutron.py:1488: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4
cloudferrylib/os/network/neutron.py:1490: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4
cloudferrylib/os/network/neutron.py:1491: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 8
cloudferrylib/os/network/neutron.py:1492: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 8
cloudferrylib/os/network/neutron.py:1493: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 8
cloudferrylib/os/network/neutron.py:1494: [W0311(bad-indentation), ] Bad indentation. Found 16 spaces, expected 12
cloudferrylib/os/network/neutron.py:1495: [W0311(bad-indentation), ] Bad indentation. Found 20 spaces, expected 16
cloudferrylib/os/network/neutron.py:173: [W0221(arguments-differ), NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:283: [W0621(redefined-outer-name), NeutronNetwork.get_network] Redefining name 'network' from outer scope (line 24)
cloudferrylib/os/network/neutron.py:759: [W0221(arguments-differ), NeutronNetwork.get_security_groups] Arguments number differs from overridden 'get_security_groups' method
cloudferrylib/os/network/neutron.py:863: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_vips] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:888: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_members] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:919: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_monitors] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:980: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_pools] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:1371: [W1504(unidiomatic-typecheck), NeutronNetwork.get_resource_hash] Using type() instead of isinstance() for a typecheck.
cloudferrylib/os/network/neutron.py:1378: [W1504(unidiomatic-typecheck), NeutronNetwork.get_resource_hash] Using type() instead of isinstance() for a typecheck.
cloudferrylib/os/network/neutron.py:1382: [W1504(unidiomatic-typecheck), NeutronNetwork.get_resource_hash] Using type() instead of isinstance() for a typecheck.

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/utils/cache.py [hunk 1]
```diff
@@ -30,7 +30,9 @@ def __init__(self, func):
         self.func = func
         self.cache = {}
 
-    def __call__(self, *args):
+    def __call__(self, *args, **kwargs):
+        if kwargs:
+            return self.func(*args, **kwargs)
```
Comment (roman-verchikov) [inline cloudferrylib/utils/cache.py]: This means cache will *never* be used for methods with key-value arguments used, please consider modifying the logic.

Comment (roman-verchikov) [inline cloudferrylib/utils/cache.py]: Also modify class' docstring accordingly

**Comment by MirantisWorkloadMobility** [comment]
Ready

File: cloudferrylib/utils/cache.py (hunk 1 above)
Comment (miarmak) [inline cloudferrylib/utils/cache.py]: As we can see in the docstring of this class, current cache implementation already does not support key-value arguments (without this commit). Moreover, it fails, if we try to use it with method, which uses key-value arguments. And for this actual patch I fixed this specific issue.

So in this case, I suggest implementing this feature (support kwargs in cache) as a separate task with it's own priority, cause it is not related to this patch.

**Comment by miarmak** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.base.network
cloudferrylib/base/network.py:25: [E0711(notimplemented-raised), Network.get_func_mac_address] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:24: [W0613(unused-argument), Network.get_func_mac_address] Unused argument 'instance'
cloudferrylib/base/network.py:28: [E0711(notimplemented-raised), Network.create_port] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'mac'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'tenant_id'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'sg_ids'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'ip'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'net_id'
cloudferrylib/base/network.py:27: [W0613(unused-argument), Network.create_port] Unused argument 'keep_ip'
cloudferrylib/base/network.py:31: [E0711(notimplemented-raised), Network.delete_port] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:30: [W0613(unused-argument), Network.delete_port] Unused argument 'port_id'
cloudferrylib/base/network.py:34: [E0711(notimplemented-raised), Network.check_existing_port] NotImplemented raised - should raise NotImplementedError
cloudferrylib/base/network.py:33: [W0613(unused-argument), Network.check_existing_port] Unused argument 'network_id'
cloudferrylib/base/network.py:33: [W0613(unused-argument), Network.check_existing_port] Unused argument 'mac'
cloudferrylib/base/network.py:37: [E0711(notimplemented-raised), Network.get_security_groups] NotImplemented raised - should raise NotImplementedError
************* Module tests.cloudferrylib.os.network.test_neutron
tests/cloudferrylib/os/network/test_neutron.py:48: [E1002(super-on-old-class), NeutronTestCase.setUp] Use of super on an old style class
tests/cloudferrylib/os/network/test_neutron.py:186: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:188: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:191: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:196: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
tests/cloudferrylib/os/network/test_neutron.py:201: [E1101(no-member), NeutronTestCase.test_get_quotas] Instance of 'NeutronTestCase' has no 'assertDictEqual' member
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:1478: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4
cloudferrylib/os/network/neutron.py:1488: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4
cloudferrylib/os/network/neutron.py:1490: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4
cloudferrylib/os/network/neutron.py:1491: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 8
cloudferrylib/os/network/neutron.py:1492: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 8
cloudferrylib/os/network/neutron.py:1493: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 8
cloudferrylib/os/network/neutron.py:1494: [W0311(bad-indentation), ] Bad indentation. Found 16 spaces, expected 12
cloudferrylib/os/network/neutron.py:1495: [W0311(bad-indentation), ] Bad indentation. Found 20 spaces, expected 16
cloudferrylib/os/network/neutron.py:173: [W0221(arguments-differ), NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:283: [W0621(redefined-outer-name), NeutronNetwork.get_network] Redefining name 'network' from outer scope (line 24)
cloudferrylib/os/network/neutron.py:759: [W0221(arguments-differ), NeutronNetwork.get_security_groups] Arguments number differs from overridden 'get_security_groups' method
cloudferrylib/os/network/neutron.py:863: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_vips] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:888: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_members] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:919: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_monitors] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:980: [W1201(logging-not-lazy), NeutronNetwork.upload_lb_pools] Specify string format arguments as logging function parameters
cloudferrylib/os/network/neutron.py:1371: [W1504(unidiomatic-typecheck), NeutronNetwork.get_resource_hash] Using type() instead of isinstance() for a typecheck.
cloudferrylib/os/network/neutron.py:1378: [W1504(unidiomatic-typecheck), NeutronNetwork.get_resource_hash] Using type() instead of isinstance() for a typecheck.
cloudferrylib/os/network/neutron.py:1382: [W1504(unidiomatic-typecheck), NeutronNetwork.get_resource_hash] Using type() instead of isinstance() for a typecheck.

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by roman-verchikov** [comment]
+1 code review, +1 from staging test team, can be merged

**Pull request merged**